### ISSUE TITLE
Status of matching

### ISSUE THREAD

**Issue by aubertc** [popbr/data-integration#9]
Please, comment on the current status of the matching problem: what is implemented, what is left to do, what issues are you facing?

**Comment by MNSleeper** [comment]
I can't get to matching right now until the foreign key/drop table issue is solved

**Comment by MNSleeper** [comment]
Yes, that is correct.

**Comment by MNSleeper** [comment]
Matching is now accessible.If you run the method, know that there are System.out.printlns I use for testing that will significantly slow the program. Be wary if you use it.

**Comment by MNSleeper** [comment]
"Running the method", in my interpretation of it, meant running the program with the Linkage methods

**Comment by MNSleeper** [comment]
After some data analysis, I can confirm that the matching program only adds items to the LinkTable if it isn't in there -- I tested this by examining a list of outputted additions to the table and confirming there are no matches. 
Notably, there are some close matches like "Loma" university and "Loma " university, and "Maryland" university and "Marlyland" University.

**Comment by aubertc** [comment]
> "Running the method", in my interpretation of it, meant running the program with the Linkage methods

So, you mean not commenting
https://github.com/popbr/data-integration/blob/15c1bf5f44cb40ea08da66ac2a1bc715f7a7533c/Project/Database-IO/src/main/java/popbr/DatabaseIO.java#L122
?

**Comment by MNSleeper** [comment]
Yes, running the methods _CreateLinkageTable_ and _LinkTable_ are what I mean

**Comment by aubertc** [comment]
Looking at https://github.com/popbr/data-integration/blob/15c1bf5f44cb40ea08da66ac2a1bc715f7a7533c/Project/Database-IO/src/main/java/popbr/DatabaseIO.java#L927 , I feel like there is a disconnection.

By "matching" you mean:

1. Parse the table(s?) created by importing the databases,
2. Create one entry in the "LinkTable" per entity (in that case, University's name)
3. Create a foreign key from from the LinkTable to the original table(s).

Am I correct? 

By "matching" I meant "during part 2., make sure that there isn't in the LinkTable an entry that already corresponds to that same entity". Of course, we can't look at this problem without doing 1./2./3. first, even partially.

**Comment by MNSleeper** [comment]
You are correct about points 1-3. and, to your next point, unique entries only, I can confirm that there are only unique entries. I confirmed this by comparing the table length of the example DB's SQL table (267 entries) versus the SQL table for the Linktable (231 entries) and I confirmed that the 36 entry deficit were all repeat entries of prior entries in the database

**Comment by MNSleeper** [comment]
Perhaps it's a misunderstanding on my part, by when I was looking up pulling all entries linked by a foreign key in SQL, i read that people use inner joins and was wondering why we don't use a large set of inner join commands over foreign keys?

Also, @aubertc , if you know a way to select all entities linked by foreign keys, I would like to know it/know where to look to get an idea.

**Comment by aubertc** [comment]
> large set of inner join commands over foreign keys?

We could probably _if_ we were not trying to identify some of those entities (i.e., try to match "Augusta Univ." with "Augusta University"). I don't think inner join can do this sort of "approximate matching", but may be wrong.

> Also, @aubertc , if you know a way to select all entities linked by a single foreign key, I would like to know it/know where to look to get an idea. It would being the vein of "Select an entity and all other entities linked to it by FKs"

Try to present the problem in a more minimal way. Write an sql script that create the databases like you want them to be, and try to write the command. I'll be happy to help if you want.

**Comment by aubertc** [comment]
> Create a foreign key from from the LinkTable to the original table(s).

Hold on, that's not right. The idea is that an entry in the LinkTable may contain an entity present in _multiple entries_ in the original data, so the foreign key should be the other way around. Or you should have a "cross-reference" table.

Does that make sense? A live discussion may be overdue on this.

**Comment by MNSleeper** [comment]
I am somewhat confused, but I do understand the base idea. A live discussion would be great.
